---
title: 'Bioinformatics and Statistical Genetic'
subtitle: 'Linkage Disequilibrium'
output:
  pdf_document:
    fig_height: 4
    toc: yes
  html_document:
    toc: yes
author: Leonardo Ortoleva & Egon Ferri
---

```{r}
knitr::opts_chunk$set(echo = F)
```



```{r message=FALSE, warning=FALSE}
require(genetics, quietly = T)
require(viridis, quietly = T)
require(knitr, quietly = T)
require(purrr, quietly = T)
require(HardyWeinberg, quietly = T)
require(data.table, quietly = T)
require(fitdistrplus, quietly = T)
require(e1071, quietly = T)
require(LDheatmap, quietly = T)
require(tidyverse, quietly = T)
require(kableExtra)
```

# 2  Load the FOXP2.dat file into the R environment. How many individuals and how many SNPs are there in the database? What percentage of the data is missing?

```{r}
data <- fread('FOXP2.dat', header = TRUE)
data <- data.frame(data)
```

```{r}
kable(data[1:10,1:7], format = "latex",  booktabs = T, caption = 'Data') %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), latex_options = 'hold_position') 
```

```{r}
nInd <- dim(data)[1]
nVar <- dim(data)[2]-1

print(paste0('The number of individuals is ', nInd))
print(paste0('The number of variants is ', nVar))
```

```{r}
napercentage=sum(is.na(data))/(length(data)*nrow(data))*100
print(paste0('Percentage of the data is missing is ', napercentage, '%'))
```

# 3 Determine the genotype counts for each SNP, and depict all SNPs simultaeneously in a ternary plot, and comment on your result. For how many variants do you reject Hardy-Weinberg equilibrium using an ordinary chi-square test without continuity correction? 

```{r}
dBim <- read.table('FOXP2.bim', header = F)
dBim<-dBim[ ,c(2, 4:6)]
colnames(dBim)=c('id','basepairpos','Allele1','Allele2')
#kable(dBim[1:10,], format = "latex",  booktabs = T) %>%
  #kable_styling(bootstrap_options = c("striped")) 
```

```{r}

dataBimOnly <- dBim[, c(3:4)]
v = c()

for(i in 1:nrow(dataBimOnly)){
  v[i] <- paste(dataBimOnly[i,1], dataBimOnly[i,2], collapse  = "/", sep = "/")
}

only_data<- data[2:ncol(data)]
only_data<-apply(only_data, 1:2, function(x) str_replace_all(x, pattern = regex("/"), replacement = regex("")))

matr <- MakeCounts(only_data, v)
colnames(matr)<-c("Homo_1","Hetero","Homo_2","NA")
kable(matr[1:10,1:3], format = "latex",  booktabs = T, caption = 'Genome counts') %>%
  kable_styling(bootstrap_options = c("striped"), latex_options = 'hold_position') 

```

```{r}
matr <- matr[, 1:3]


HWTernaryPlot(matr, cex = 0.25, pch=2, vertexlab = c("Homo_1","Hetero","Homo_2"))
 

chiSqOut <- HWChisqStats(matr, pvalues = T)

signi<-sum(chiSqOut<0.05)
print(paste0("The number of variants SNPs for which we reject HWE is: ", signi))

perc<- signi*100/length(chiSqOut)
print(paste0('Percentage of out of equilibrium SNPs is: ', round(perc,3), '%'))
```

From the ternary plot we see that almost every SNPs is in equilibrium, as we see from the $\chi^2$ in the $94$% of the cases we accept the null hypotesis of being in equilibrium. There is still a $6$% of cases where we reject the null hypotesis, but as the plot shows they are still very near to the confidence bands.

Notes about notation: We defined as Homo_1 and Homo_2 the homozigotes genotypes (since we do not have always 'AA' and 'BB') and as Hetero the heterozygote. The graph is unbalanced to the right because (we guess) in the .bim file the rarest allele is displayed before.

# 4 Using the function LD from the genetics package, compute the LD statistic D for the SNPs rs34684677 and rs2894715 of the database. Is there significant association between the alleles of these two SNPs?

```{r}
SNP1 <- genotype(data$rs34684677)
SNP2<- genotype(data$rs2894715)


res1 <- LD(SNP1,SNP2)
res1
```


We see that we have a absolute deviation from the indipendence of $0.055$, that once standardized become a $D'$ of almost one, meaning that the two SNPs are coinherited almost always. So the p-value is very small (of the order of $10^{-6}$) and we can strongly reject the null hypotesis of the two SNPs being indipendent. 

# 5 Also compute the LD statistic D for the SNPs rs34684677 and rs998302 of the database. Is there significant association between these two SNPs? Is there any reason why rs998302 could have stronger or weaker correlation than rs2894715?

```{r}

SNP3<- genotype(data$rs998302)


res2 <- LD(SNP1,SNP3)
res2
```


We see that we have a absolute deviation from the indipendence of $0.007$, that once standardized become a $D'$ of $0.179$. The p-value is $0.189$ and we can't reject the null hypotesis of the two SNPs being indipendent, even with a not-so-strict alpha of $0.1$.

```{r}
kable(head(data[,c('rs2894715','rs34684677', 'rs998302') ]), format = "latex",  booktabs = T, caption = 'Genotypes') %>%
  kable_styling(bootstrap_options = c("striped"), latex_options = 'hold_position') 
```

```{r}
print('positions of genomes rs34684677, rs2894715, rs998302:')
print(paste0("columns in our dataset: ",which(colnames(data)=="rs34684677"),", ",which(colnames(data)=="rs2894715"), ", " ,which(colnames(data)=='rs998302')))

print(paste0("basepair: ", dBim[dBim['id']=="rs34684677"][2],", ",dBim[dBim['id']=="rs2894715"][2], ", " ,dBim[dBim['id']=='rs998302'][2]))
```

Genetic variants that are physically close on a chromosome typically have high correlations, so we can suppose this is why rs998302  has  weaker correlation than rs2894715.


# 6 Given your previous estimate of D for SNPs rs34684677 and rs2894715, infer the haplotype frequencies. Which haplotype is the most common?

$$D = p_{AB} - p_Ap_B$$
pA and pB can be estimated by the sample allele frequencies $\hat p_A$ and $\hat p_B$

$$p_{AB} = D+p_Ap_B$$
```{r}
res3<-summary(SNP1)
res4<-summary(SNP2)
```

```{r}
D<-res1$D
G_<-as.numeric(res3$allele.freq[,2][1])
T_<-as.numeric(res3$allele.freq[,2][2])
t_<-as.numeric(res4$allele.freq[,2][1])
g_<-as.numeric(res4$allele.freq[,2][2])


c1<-(G_*t_)+D
c2<-(G_*g_)-D
c3<-(T_*t_)-D
c4<-(T_*g_)+D

DF<-data.frame(row.names = c('G', 'T'))
DF['T']<-c(c1, c3)
DF['G']<-c(c2,c4)
kable(DF, format = "latex",  booktabs = T) %>%
  kable_styling(bootstrap_options = c("striped"), latex_options = 'hold_position') %>%
  add_header_above(c("rs34684677","rs2894715" = 2))

```

So the most common haplotype is G T (G from $rs34684677$ and T from $rs2894715$).


# 7 Compute the LD statistics R2 for all the marker pairs in this data base, using the LD function of the packages genetics. Also compute an alternative estimate of R2 obtained by using the PLINK program. Make a scatter plot for R’s LD estimates against PLINK’s LD estimates. Are they identical or do they at least correlate? What’s the difference between these two estimators? Which estimator would your prefer and why?

```{r}
RES <- data.frame(genotype(only_data[,1],sep=""))
for(i in 2:ncol(only_data)) {
   snp <- genotype(only_data[,i],sep="")
   RES <- cbind(RES,snp)
}

```




```{r}
#output <- LD(RES)
```



```{r}
#saveRDS(output, "output.rds")
output <- readRDS("output.rds")
```





```{r}
R2 <- output$"R^2"
R2 <- R2[upper.tri(R2)]
print('LD function R^2 estimates using package genetics')
R2[1:20]
```

```{r}
fromplink<-data.frame(fread('FOXP2.ld'))
fromplink <- fromplink[upper.tri(fromplink, diag = F)]
print('LD function R^2 estimates using plink')
fromplink[1:20]
```

```{r}
plot(R2, fromplink, col='lightsalmon',
     main = title("Scatterplot of the two estimation"), xlab = "R^2 estimated with r", 
     ylab= "R^2 estimated with plink")
abline(a=0, b=1, lwd=2)
```

```{r}
correlation<-cor(R2, fromplink)
print(paste0('The correlation coefficient is: ', round(correlation,2)))
```

They are not identical but actually  they do correlate almost perfectly.

The first estimator evaluates precisely the LD statistic of each pair, but takes a lot of time.
The plink estimator is almost immediate so we can suppose that there is some optimization that allows us to reduce the calculation time, but as we can see the cost is almost 0 because the statistics are almost the same, due to this, we can say that we prefer the second estimator.

# 8 Compute a distance matrix with the distance in base pairs between all possible pairs of SNPs, using the basepair position of each SNP given in the .bim file. Make a plot of R’s R2 statistics against the distance (expressed as the number of basepairs) between the markers. Comment on your results.

```{r}
dij <- dist(dBim[, 2])
plot(R2, dij, col='lightsalmon', pch=3, ylab = 'distances in basepairs between the two markers')
segments(0,300000,1,0, lwd=2, lty=2)
```

Genetic variants that are physically close on a chromosome typically have high correlations (as we can see from the plot) but this is not a strong correlation since we have a lot of values out of our expectations; especially we have a lot of close pairs with weak correlation.
As we will see in next paragraphs, pruning the SNPs with small minor allele frequencies can help in reducing this behaviour.

# 9 Make an LD heatmap of the markers in this database, using the R2 statistic with the LD function. Make another heatmap obtained by filtering out all variants with a MAF below 0.35, and redoing the computations to obtain the R2 statistics in R. Can you explain any differences observed between the two heatmaps?


```{r}

LDheatmap(RES,LDmeasure="r",color=viridis(18))

```

```{r}
MAF<- function(snp){
  res<- summary(genotype(snp, sep=''))
  return(min(res$allele.freq))
}

mafs<- apply(only_data, 2, MAF)
mafs_big<-only_data[,which(mafs>0.35)]
```

```{r}
RES_big <- data.frame(genotype(mafs_big[,1],sep=""))
for(i in 2:ncol(mafs_big)) {
   snp <- genotype(mafs_big[,i],sep="")
   RES_big <- cbind(RES_big,snp)
}
```

```{r}
LDheatmap(RES_big,LDmeasure="r",color=viridis(18))
```

The difference is clear: flitering out with small MAFs the blocks that we can barely see in the first plot become more visible since the correlation are a lot stronger.

# 10 Can you distinguish blocks of correlated markers in the area of the FOXP2 gene? How many blocks do you think that at least seem to exist?

We can see two enormous blocks with an $r$ of minimum 0.4 that cut the data in two clusters.
In the first cluster (up-right) we see two really strong blocks. In the other big cluster we can see three smaller strong blocks.


# 11 Simulate independent SNPs under the assumption of Hardy-Weinberg equilibrium. Simulate as many SNPs as you have in your database, and take care to match each SNP in your database with a simulated SNP that has the same sample size and allele frequency. Make an LD heatmap of the simulated SNPs, using R2 as your statistic. Compare the results with the LD heatmap of the FOXP2 region. What do you observe? State your conclusions

```{r}
simulator<-function(geno){
  geno<-summary(genotype(geno, sep=''))
  p<-as.numeric(geno$allele.freq[,2][1])
  q<-as.numeric(geno$allele.freq[,2][2])
  simul<-sample(c("AA","AB","BB"),104,replace=TRUE,prob=c(p*p,2*p*q,q*q))
  return(simul)
}

result<- apply(FUN = simulator,X= only_data,MARGIN = 2)
```

```{r}
RES_simul <- data.frame(genotype(result[,1],sep=""))
for(i in 2:ncol(result)) {
   snp <- genotype(result[,i],sep="")
   RES_simul <- cbind(RES_simul,snp)
}
```

```{r}
LDheatmap(RES_simul,LDmeasure="r",color=viridis(18))
```

As we could expect, simulating this way we don't have any correlation between variables, since we for construcion we are simulating each SNPs for each individual independently.